### DISCUSSION TITLE
AVR compatibilies macros

### DISCUSSION THREAD

**Discussion by PontusO** [earlephilhower/arduino-pico#2248] [General]
In Arduino.h we have the following AVR compatibility macros:

#define portOutputRegister(port)    ((volatile uint32_t*) sio_hw->gpio_out)
#define portInputRegister(port)     ((volatile uint32_t*) sio_hw->gpio_in)
#define portModeRegister(port)      ((volatile uint32_t*) sio_hw->gpio_oe)

I am wondering if these were not meant to be:

#define portOutputRegister(port)    ((volatile uint32_t *)&(sio_hw->gpio_out))
#define portInputRegister(port)     ((volatile uint32_t *)&(sio_hw->gpio_in))
#define portModeRegister(port)      ((volatile uint32_t *)&(sio_hw->gpio_oe))

instead !?
I am porting an old application that runs on multiple other targets but with the pico BSP these fail.

**Comment by earlephilhower** [comment]
Yeah, that looks like a bug you found.  The defines should include the `&` otherwise they'll use the value of the register as a pointer instead an bad things will happen.

Want to do the PR for it?

**Reply by PontusO** [reply]
Bad things did indeed happen =) Yeah, I'll create a PR for it.
